features/call/impl/src/main/kotlin/io/element/android/features/call/impl/ui/ElementCallActivity.kt

```diff
@@ -185,23 +188,23 @@ class ElementCallActivity :
 
     private fun setCallType(intent: Intent?) {
         val callType = intent?.let {
-            IntentCompat.getParcelableExtra(it, DefaultElementCallEntryPoint.EXTRA_CALL_TYPE, CallType::class.java)
+            IntentCompat.getParcelableExtra(intent, DefaultElementCallEntryPoint.EXTRA_CALL_TYPE, CallType::class.java)
+                ?: intent.dataString?.let(::parseUrl)?.let(::ExternalUrl)
         }
-        val intentUrl = intent?.dataString?.let(::parseUrl)
-        when {
-            // Re-opened the activity but we have no url to load or a cached one, finish the activity
-            intent?.dataString == null && callType == null && webViewTarget.value == null -> finish()
-            callType != null -> {
-                webViewTarget.value = callType
-                presenter = presenterFactory.create(callType, this)
-            }
-            intentUrl != null -> {
-                val fallbackInputs = CallType.ExternalUrl(intentUrl)
-                webViewTarget.value = fallbackInputs
-                presenter = presenterFactory.create(fallbackInputs, this)
-            }
-            // Coming back from notification, do nothing
-            else -> return
+        val currentCallType = webViewTarget.value
+        if (currentCallType == null && callType == null) {
+            Timber.tag(loggerTag).d("Re-opened the activity but we have no url to load or a cached one, finish the activity")
+            finish()
+        } else if (currentCallType == null) {
+            Timber.tag(loggerTag).d("Set the call type and create the presenter")
+            webViewTarget.value = callType
+            presenter = presenterFactory.create(callType!!, this)
```

Comment:
Sadly i had to use `!!` here, the compiler is not clever enough (or is it me?)

Reply:
You could also explicitly add `&& callType != null` to the if clause.

Reply:
Yes, but `callType != null` is always `true` in this case, so I think it's worse.